propolis/src/block.rs

```diff
@@ -53,53 +54,138 @@ pub struct BlockInquiry {
 pub trait BlockDev<R: BlockReq>: Send + Sync + 'static {
     /// Enqueues a [`BlockReq`] to the underlying device.
     fn enqueue(&self, req: R);
+
     /// Requests metadata about the block device.
     fn inquire(&self) -> BlockInquiry;
 }
 
+/// Abstraction over an actual backing store
+pub trait BlockDevBackingStore: Send + Sync + 'static {
+    fn issue_read(&self, offset: usize, sz: usize) -> Result<Vec<u8>>;
+    fn issue_write(&self, bytes: &[u8], offset: usize) -> Result<()>;
+    fn issue_flush(&self) -> Result<()>;
+
+    fn is_ro(&self) -> bool;
+
+    // length in bytes
+    fn len(&self) -> usize;
+}
+
 /// A block device implementation backed by an underlying file.
-///
-/// Primarily accessed through the [`BlockDev`] interface.
-pub struct PlainBdev<R: BlockReq> {
+pub struct FileBlockDevBackingStore {
     fp: File,
     is_ro: bool,
-    is_raw: bool,
-    block_size: usize,
-    sectors: usize,
-    reqs: Mutex<VecDeque<R>>,
-    cond: Condvar,
+    len: usize,
 }
 
-impl<R: BlockReq> PlainBdev<R> {
-    /// Creates a new block device from a device at `path`.
-    pub fn create(path: impl AsRef<Path>, readonly: bool) -> Result<Arc<Self>> {
+impl FileBlockDevBackingStore {
+    pub fn from_path(
+        path: impl AsRef<Path>,
+        readonly: bool,
+    ) -> Result<FileBlockDevBackingStore> {
         let p: &Path = path.as_ref();
 
         let meta = metadata(p)?;
         let is_ro = readonly || meta.permissions().readonly();
 
         let fp = OpenOptions::new().read(true).write(!is_ro).open(p)?;
-        let is_raw = fp.metadata()?.file_type().is_char_device();
+        let is_ro = readonly || meta.permissions().readonly();
+        let len = fp.metadata().unwrap().len() as usize;
+
+        Ok(FileBlockDevBackingStore { fp: fp, is_ro: is_ro, len: len })
+    }
+
+    pub fn get_file(&self) -> &File {
+        &self.fp
+    }
+}
+
+impl BlockDevBackingStore for FileBlockDevBackingStore {
+    fn issue_read(&self, offset: usize, sz: usize) -> Result<Vec<u8>> {
+        let mut vec = Vec::<u8>::with_capacity(sz);
```

Comment:
A bit unfortunate that we need to allocate an intermediate buffer here vs the previous behaviour of reading directly into the mapped guest memory. Could we just modify `issue_read` (and similarly `issue_write`) to be passed in a `GuestRegion`?

Reply:
Good idea, this makes it much cleaner. Fixed in 6b79bfe.